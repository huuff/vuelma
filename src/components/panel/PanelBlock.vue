<template>
<slot></slot>
</template>

<script setup lang="ts">
import { Component } from "vue";
import { OptionalId } from "@/util/optional-id";
import { OptionalIcon } from '@/util/optional-icon';
import { FontAwesomeIconName } from '@/types/fontawesome-icon-name';

export interface PanelBlockProps extends OptionalId, OptionalIcon {
  titleText: string;
  itemId?: string;
  icon?: FontAwesomeIconName;
  tag?: string | Component;
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const props = defineProps<PanelBlockProps>();
</script>
