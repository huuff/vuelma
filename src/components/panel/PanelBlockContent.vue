<template>
<slot></slot>
</template>

<script setup lang="ts">
// This is just a wrapper for block content

export type PanelBlockContentProps = {
  tag?: 'div' | 'label';
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const props = defineProps<PanelBlockContentProps>();
</script>
