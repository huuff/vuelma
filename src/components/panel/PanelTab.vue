<template>
<slot></slot>
</template>

<script setup lang="ts">
import { OptionalId } from '@/util/optional-id';

export interface PanelTabProps extends OptionalId {
  titleText: string;
  itemId?: string;
}

const props = defineProps<PanelTabProps>();
</script>
