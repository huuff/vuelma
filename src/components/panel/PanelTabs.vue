<template>
<slot></slot>
</template>

<script setup lang="ts">
// This component is only for grouping tabs
</script>
