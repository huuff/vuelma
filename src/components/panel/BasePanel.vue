<template>
<nav class="panel">
  <p v-if="headingText" class="panel-heading"> {{ headingText }} </p>
  <render-content />
</nav>
</template>

<script setup lang="tsx">
// TODO: Tabs
// TODO: Icons
// TODO: Content blocks
// TODO: Label blocks
import { useSlots, toRef, VNode } from "vue";
import { useOptionalTwoWayBinding } from '@/composables/optional-two-way-binding';
import { unwrapFragment } from "@/util/unwrap-fragment";
import PanelBlock, { PanelBlockProps } from './PanelBlock.vue';
import { getId } from '@/util/optional-id';
import partial from "lodash/partial";
import classnames from "classnames";

const props = defineProps<{
  headingText?: string;
  activeItemId?: string;
  activeTabId?: string;
}>();

const emit = defineEmits<{
  (event: "update:activeItemId", itemId?: string): void;
  (event: "update:activeTabId", tabId?: string): void;
}>();

// eslint-disable-next-line
// @ts-ignore
const actualActiveItemId = useOptionalTwoWayBinding(undefined, toRef(props, "activeItemId"), partial(emit, "update:activeItemId"));

// eslint-disable-next-line
// @ts-ignore
const actualActiveTabId = useOptionalTwoWayBinding(undefined, toRef(props, "activeTabId"), partial(emit, "update:activeTabId"));

function renderBlock(node: VNode): VNode {
  const blockProps = node.props as PanelBlockProps;
  const id = getId(blockProps);

  return <a
            class={classnames({
              "panel-block": true,
              "is-active": id === actualActiveItemId.value,
            })}
            onClick={() => actualActiveItemId.value = id}
        > { blockProps.titleText }
        </a>
            
}

const slots = useSlots();

const renderContent = () => slots.default && unwrapFragment(slots.default()).map(node => {
  if (node.type === PanelBlock) 
    return renderBlock(node);
});
</script>
