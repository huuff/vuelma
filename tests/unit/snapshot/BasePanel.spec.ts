import { mount } from "@vue/test-utils";
import BasePanel from "@/components/panel/BasePanel.vue";
import PanelBlock from "@/components/panel/PanelBlock.vue";
import PanelTabs from "@/components/panel/PanelTabs.vue";
import PanelTab from "@/components/panel/PanelTab.vue";
import PanelBlockContent from "@/components/panel/PanelBlockContent.vue";

test("BasePanel.vue snapshot", () => {
  const wrapper = mount(BasePanel, {
    props: {
      headingText: "Heading text",
      activeItemId: "Item 1",
      activeTabId: "Tab 1",
      color: "warning",
    },
    global: {
      components:{ PanelBlock, PanelTabs, PanelTab, PanelBlockContent },
    },
    slots: {
      default: `
        <panel-block titleText="Item 1" />
        <panel-tabs>
          <panel-tab titleText="Tab 1">
            <panel-block titleText="Item 2" />
            <panel-block titleText="Item 3" />
            <panel-block titleText="Item 4" />
          </panel-tab>
          <panel-tab titleText="Tab 2">
            <panel-block titleText="Item 5" />
            <panel-block-content>
              <label>
                <input type="text" /> Test
              </label>
            </panel-block-content>
          </panel-tab>
        </panel-tabs>
      `
    },
  });

  expect(wrapper.element).toMatchSnapshot();
});
