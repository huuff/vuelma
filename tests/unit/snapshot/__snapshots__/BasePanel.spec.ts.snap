// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasePanel.vue snapshot 1`] = `
<nav
  class="panel is-warning"
>
  <p
    class="panel-heading"
  >
    Heading text
  </p>
  
  <a
    class="panel-block is-active"
    titletext="Item 1"
  >
    <span>
      Item 1
    </span>
  </a>
  
  <p
    class="panel-tabs"
  >
    <a
      class="is-active"
      titletext="Tab 1"
    >
      Tab 1
    </a>
    <a
      class=""
      titletext="Tab 2"
    >
      Tab 2
    </a>
  </p>
  <a
    class="panel-block"
    titletext="Item 2"
  >
    <span>
      Item 2
    </span>
  </a>
  <a
    class="panel-block"
    titletext="Item 3"
  >
    <span>
      Item 3
    </span>
  </a>
  <a
    class="panel-block"
    titletext="Item 4"
  >
    <span>
      Item 4
    </span>
  </a>
  
  
</nav>
`;
